Generated binding compile time error when conforming to 2 protocols with the same @required member

Since two protocols have the same required member we generate it twice - so the C# compiler is not happy.

AFAIK the two protocols can be used separately, so the definition should be in both, or together (and the declaration should be merged)

### Steps to Reproduce

1. Uncomment `bool Available { [Bind ("isAvailable")] get; }` in `NFCNdefTag` (corennfc.cs)
2. Build

### Expected Behavior

No error.

### Actual Behavior

```
error CS0102: The type 'NFCFeliCaTagWrapper' already contains a definition for 'Available'
```

### Environment

`xcode11`

### Example Project (If Possible)

See (upcoming) PR for `corenfc` support

--- a/docs/website/generator-errors.md
+++ b/docs/website/generator-errors.md
@@ -196,6 +196,14 @@ This usually indicates a bug in Xamarin.iOS/Xamarin.Mac; please [file a bug repo
 
 ### <a name='BI1066'/>BI1066: Unsupported return type '{type}' in {method}.
 
+### <a name='BI1067'/>BI1067: The type '{type}' is trying to inline the property '{property}' from the protocols '{readonly.DeclaringType.FullName}' and '{writeonly.DeclaringType.FullName}', but the inlined properties don't share the same accessors ('{readonly}' is read-only, while '${writeonly}' is write-only).
+
+### <a name='BI1068'/>BI1068: The type '{type}' is trying to inline the property '{property}' from the protocols '{protocol1}' and '{protocol2}', and the inlined properties use different selectors ({protocol1} uses '{selector1}', and {protocol2} uses '{selector2}.
+
+### <a name='BI1069'/>BI1069: The type '{type}' is trying to inline the methods binding the selector '{selector}' from the protocols '{protocol1}' and '{protocol2}', using methods with different signatures ('{signature}' vs '{signature2}').
+
+### <a name='BI1070'/>BI1070: The type '{type}' is trying to inline the property '{property}' from the protocols '{protocol1}' and '{protocol2}', but the inlined properties are of different types ('{property1}' is int, while '{property2}' is int).
+
 # BI11xx: warnings
 
 <!-- 11xx: warnings -->

--- a/src/corenfc.cs
+++ b/src/corenfc.cs
@@ -652,11 +652,9 @@ namespace CoreNFC {
 	[Protocol (Name="NFCNDEFTag")]
 	interface NFCNdefTag : NSSecureCoding, NSCopying {
 
-		// FIXME https://github.com/xamarin/xamarin-macios/issues/6366
-		// error CS0102: The type 'NFCFeliCaTagWrapper' already contains a definition for 'Available'
-		// [Abstract]
-		// [Export ("available")]
-		// bool Available { [Bind ("isAvailable")] get; }
+		[Abstract]
+		[Export ("available")]
+		bool Available { [Bind ("isAvailable")] get; }
 
 		[Abstract]
 		[Export ("queryNDEFStatusWithCompletionHandler:")]

--- a/src/generator.cs
+++ b/src/generator.cs
@@ -5806,8 +5806,104 @@ public partial class Generator : IMemberGatherer {
 		print ("}");
 		print ("");
 		// Methods
+		// First find duplicates and select the best one. We use the selector to determine what's a duplicate.
+		var methodData = requiredInstanceMethods.Select ((v) => {
+			return new MethodData {
+				Method = v,
+				ExportAttribute = GetExportAttribute (v),
+			};
+		}).ToArray ();
+		var methodsGroupedBySelector = methodData.GroupBy ((v) => v.ExportAttribute.Selector);
+		var duplicateMethodsGroupedBySelector = methodsGroupedBySelector.Where ((v) => v.Count () > 1);
+		if (duplicateMethodsGroupedBySelector.Any ()) {
+			// Yes, there are multiple methods with the same selector. Now we must compute the signature for all of them.
+			var computeSignature = new Func<MethodInfo, string> ((MethodInfo minfo) => {
+				var sig = new StringBuilder ();
+				sig.Append (minfo.Name).Append (" (");
+				foreach (var param in minfo.GetParameters ())
+					sig.Append (param.ParameterType.FullName).Append (' ');
+				sig.Append (')');
+				return sig.ToString ();
+			});
+			foreach (var gr in duplicateMethodsGroupedBySelector)
+				foreach (var m in gr)
+					m.Signature = computeSignature (m.Method);
+
+			// Verify that all the duplicate methods have the same signature.
+			// If not, show an error, and if they all have the same signature, remove all but one from the list of methods to generate.
+			foreach (var gr in duplicateMethodsGroupedBySelector) {
+				var distinctMethodsBySignature = gr.GroupBy ((v) => v.Signature).Select ((v) => v.First ()).ToArray ();
+				if (distinctMethodsBySignature.Length > 1) {
+					exceptions.Add (ErrorHelper.CreateError (1069, $"The type '{type.FullName}' is trying to inline the methods binding the selector '{gr.Key}' from the protocols '{distinctMethodsBySignature [0].Method.DeclaringType.FullName}' and '{distinctMethodsBySignature [1].Method.DeclaringType.FullName}'," +
+$" using methods with different signatures ('{distinctMethodsBySignature [0].Method.ToString ()}' vs '{distinctMethodsBySignature [1].Method.ToString ()}')."));
+					continue;
+				}
+
+				// Remove all but the first method from the list required instance methods
+				var exclude = gr.Skip (1).Select ((v) => v.Method).ToArray ();
+				requiredInstanceMethods.RemoveAll ((v) => exclude.Contains (v));
+			}
+
+		}
+		// Go generate code for the methods
 		foreach (var mi in requiredInstanceMethods) {
 			GenerateMethod (type, mi, false, null, false, true, isBaseWrapperProtocolMethod:true);
+		}
+
+		// Properties
+		// First find duplicates and select the best one. Here we use the property name to find duplicates.
+		var propertyDuplicates = requiredInstanceProperties.GroupBy ((v) => v.Name).Where ((v) => v.Count () > 1).ToArray ();
+		if (propertyDuplicates.Any ()) {
+			foreach (var gr in propertyDuplicates) {
+				// Check that all the selectors in the group are identical
+				var properties = gr.ToArray ();
+				var exportAttributes = new ExportAttribute [properties.Length];
+				PropertyInfo readwrite = null;
+				PropertyInfo @readonly = null;
+				PropertyInfo @writeonly = null;
+				for (var i = 0; i < properties.Length; i++) {
+					// Check that all the duplicates use the same selector, and if not show a warning.
+					exportAttributes [i] = GetExportAttribute (properties [i]);
+					if (i > 0 && exportAttributes [i].Selector != exportAttributes [0].Selector) {
+						ErrorHelper.Warning (1068, $"The type '{type.FullName}' is trying to inline the property '{gr.Key}' from the protocols '{properties [0].DeclaringType.FullName}' and '{properties [i].DeclaringType.FullName}'," +
+							$" and the inlined properties use different selectors ({properties [0].DeclaringType.Name}.{properties [0].Name} uses '{exportAttributes [0].Selector}', and {properties [i].DeclaringType.Name}.{properties [i].Name} uses '{exportAttributes [i].Selector}'.");
+					}
+					if (properties [i].CanRead && properties [i].CanWrite) {
+						readwrite = properties [i];
+					} else if (!properties [i].CanRead) {
+						writeonly = properties [i];
+					} else if (!properties [i].CanWrite) {
+						@readonly = properties [i];
+					}
+				}
+
+				// Check that all the duplicates have the same property type
+				var propertyTypes = properties.GroupBy ((v) => v.PropertyType.FullName).Select ((v) => v.First ()).ToArray ();
+				if (propertyTypes.Length > 1) {
+					exceptions.Add (ErrorHelper.CreateError (1070, $"The type '{type.FullName}' is trying to inline the property '{gr.Key}' from the protocols '{propertyTypes [0].DeclaringType.FullName}' and '{propertyTypes [1].DeclaringType.FullName}'," +
+						$" but the inlined properties are of different types ('{propertyTypes [0]}' is {FormatType (type, propertyTypes [0].PropertyType)}, while '{propertyTypes [1]}' is {FormatType (type, propertyTypes [0].PropertyType)})."));
+				}
+
+				// Select the best match of the properties: prefer a read/write property if it exists, otherwise a readonly or writeonly property.
+				PropertyInfo bestMatch;
+				if (readwrite != null) {
+					bestMatch = readwrite;
+				} else if (@readonly != null && writeonly != null) {
+					exceptions.Add (ErrorHelper.CreateError (1067, $"The type '{type.FullName}' is trying to inline the property '{gr.Key}' from the protocols '{@readonly.DeclaringType.FullName}' and '{writeonly.DeclaringType.FullName}'," +
+						$" but the inlined properties don't share the same accessors ('{@readonly}' is read-only, while '${writeonly}' is write-only)."));
+					continue;
+				} else if (@readonly != null) {
+					bestMatch = @readonly;
+				} else if (writeonly != null) {
+					bestMatch = writeonly;
+				} else {
+					exceptions.Add (ErrorHelper.CreateError (99, $"Internal error: property {properties [0]} doesn't have neither a getter nor a setter."));
+					continue;
+				}
+				// Finally remove the properties we don't want to generate.
+				var exclude = properties.Where ((v) => v != bestMatch);
+				requiredInstanceProperties.RemoveAll ((v) => exclude.Contains (v));
+			}
 		}
 		foreach (var pi in requiredInstanceProperties) {
 			GenerateProperty (type, pi, null, false, true);
@@ -5819,6 +5915,13 @@ public partial class Generator : IMemberGatherer {
 			indent--;
 			print ("}");
 		}
+	}
+
+	class MethodData
+	{
+		public MethodInfo Method;
+		public ExportAttribute ExportAttribute;
+		public string Signature;
 	}
 
 	bool ConformToNSCoding (Type type)

--- a/tests/generator/BGenTests.cs
+++ b/tests/generator/BGenTests.cs
@@ -604,6 +604,12 @@ namespace GeneratorTests
 			BuildFile (Profile.iOS, "tests/vsts-970507.cs");
 		}
 
+		[Test]
+		public void DiamondProtocol ()
+		{
+			BuildFile (Profile.iOS, "tests/diamond-protocol.cs");
+		}
+
 		BGenTool BuildFile (Profile profile, params string [] filenames)
 		{
 			return BuildFile (profile, true, false, filenames);

--- a/tests/generator/ErrorTests.cs
+++ b/tests/generator/ErrorTests.cs
@@ -695,6 +695,22 @@ namespace BI1066Errors
 		}
 
 		[Test]
+		public void BI1067_1070 ()
+		{
+			BGenTool bgen = new BGenTool {
+				Profile = Profile.iOS,
+			};
+			bgen.CreateTemporaryBinding (File.ReadAllText (Path.Combine (Configuration.SourceRoot, "tests", "generator", "tests", "diamond-protocol-errors.cs")));
+			bgen.AssertExecuteError ("build");
+			bgen.AssertError (1067, "The type 'DiamondProtocol.A.C' is trying to inline the property 'P1' from the protocols 'DiamondProtocol.A.P1' and 'DiamondProtocol.A.P2', but the inlined properties don't share the same accessors ('DiamondProtocol.A.P1 P1' is read-only, while '$DiamondProtocol.A.P2 P1' is write-only).");
+			bgen.AssertWarning (1068, "The type 'DiamondProtocol.D.C' is trying to inline the property 'P1' from the protocols 'DiamondProtocol.D.P1' and 'DiamondProtocol.D.P2', and the inlined properties use different selectors (P1.P1 uses 'pA', and P2.P1 uses 'pB'.");
+			bgen.AssertError (1069, "The type 'DiamondProtocol.Y.C' is trying to inline the methods binding the selector 'm1:' from the protocols 'DiamondProtocol.Y.P1' and 'DiamondProtocol.Y.P2', using methods with different signatures ('Void M1(System.Int32)' vs 'Int32 M1(System.Boolean)').");
+			bgen.AssertError (1070, "The type 'DiamondProtocol.C.C' is trying to inline the property 'P1' from the protocols 'DiamondProtocol.C.P1' and 'DiamondProtocol.C.P2', but the inlined properties are of different types ('DiamondProtocol.C.P1 P1' is int, while 'DiamondProtocol.C.P2 P1' is int).");
+			bgen.AssertErrorCount (3);
+			bgen.AssertWarningCount (1);
+		}
+
+		[Test]
 		[TestCase (Profile.iOS)]
 		[TestCase (Profile.macOSFull)]
 		[TestCase (Profile.macOSMobile)]

--- a/tests/generator/generator-tests.csproj
+++ b/tests/generator/generator-tests.csproj
@@ -73,6 +73,8 @@
     <None Include="tests\ref-out-parameters.cs" />
     <None Include="tests\return-release.cs" />
     <None Include="tests\vsts-970507.cs" />
+    <None Include="tests\diamond-protocol.cs" />
+    <None Include="tests\diamond-protocol-errors.cs" />
   </ItemGroup>
   <ItemGroup>
     <Folder Include="tests\" />

--- a/tests/generator/tests/diamond-protocol-errors.cs
+++ b/tests/generator/tests/diamond-protocol-errors.cs
@@ -1,0 +1,154 @@
+using System;
+using Foundation;
+
+namespace DiamondProtocol
+{
+	namespace A
+	{
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("p1")]
+			int P1 { get; }
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("p1")]
+			int P1 { set; }
+		}
+
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+
+	namespace B
+	{
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("p1")]
+			int P1 { get; }
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("p1")]
+			int P1 { get; set; }
+		}
+
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+
+	namespace C
+	{
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("p1")]
+			int P1 { get; set; }
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("p1")]
+			long P1 { get; set; }
+		}
+
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+
+	namespace D
+	{
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("pA")]
+			int P1 { get; set; }
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("pB")]
+			int P1 { get; set; }
+		}
+
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+	namespace Z
+	{
+
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("m1")]
+			void M1 ();
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("m1")]
+			int M1 ();
+		}
+
+
+		//[BaseType (typeof (NSObject))]
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+
+	namespace Y
+	{
+
+		[Protocol]
+		interface P1
+		{
+			[Abstract]
+			[Export ("m1:")]
+			void M1 (int why);
+		}
+
+		[Protocol]
+		interface P2
+		{
+			[Abstract]
+			[Export ("m1:")]
+			int M1 (bool because);
+		}
+
+
+		//[BaseType (typeof (NSObject))]
+		[Protocol]
+		interface C : P1, P2
+		{
+		}
+	}
+}

--- a/tests/generator/tests/diamond-protocol.cs
+++ b/tests/generator/tests/diamond-protocol.cs
@@ -1,0 +1,64 @@
+using System;
+using Foundation;
+
+namespace DiamondProtocol
+{
+	[Protocol]
+	interface P1
+	{
+		[Abstract]
+		[Export ("p1")]
+		int P1 { get; }
+
+		[Abstract]
+		[Export ("p2")]
+		int P2 { set; }
+
+		[Abstract]
+		[Export ("p3")]
+		int P3 { get; set; }
+
+		[Abstract]
+		[Export ("m1")]
+		void M1 ();
+
+		[Abstract]
+		[Export ("m2:")]
+		void M2 (int foo);
+	}
+
+	[Protocol]
+	interface P2
+	{
+		[Abstract]
+		[Export ("p1")]
+		int P1 { get; }
+
+		[Abstract]
+		[Export ("p2")]
+		int P2 { set; }
+
+		[Abstract]
+		[Export ("p3")]
+		int P3 { get; set; }
+
+		[Abstract]
+		[Export ("m1")]
+		void M1 ();
+
+		[Abstract]
+		[Export ("m2:")]
+		void M2 (int foo);
+
+		[Abstract]
+		[Export ("m3:")]
+		void M3 (string foo);
+	}
+
+
+	//[BaseType (typeof (NSObject))]
+	[Protocol]
+	interface C : P1, P2
+	{
+	}
+}

--- a/tests/xtro-sharpie/iOS-CoreNFC.ignore
+++ b/tests/xtro-sharpie/iOS-CoreNFC.ignore
@@ -1,6 +1,3 @@
-## generator issue -> https://github.com/xamarin/xamarin-macios/issues/6366
-!missing-protocol-member! NFCNDEFTag::isAvailable not found
-
 ## deprecated/replaced selector (typo) - will be removed in a future beta
 !missing-selector! +NFCNDEFPayload::wellKnowTypeTextPayloadWithString:locale: not bound
 

